shapely/set_operations.py

```diff
@@ -501,3 +503,110 @@ def coverage_union_all(geometries, axis=None, **kwargs):
     # create_collection acts on the inner axis
     collections = lib.create_collection(geometries, GeometryType.GEOMETRYCOLLECTION)
     return lib.coverage_union(collections, **kwargs)
+
+
+@multithreading_enabled
+def disjoint_subset_union(a, b, **kwargs):
+    """Merges multiple polygons into one. This is an optimized version of
+    union which assumes inputs can be divided into subsets that do not intersect.
+
+    If there is only one such subset, performance can be expected to be worse than
+    :func:`union`. As such, it is recommeded to use ``disjoint_subset_union`` with
+    GeometryCollections rather than individual geometries.
+
+    Requires GEOS >= 3.12.0
+
+    Parameters
+    ----------
+    a : Geometry or array_like
+    b : Geometry or array_like
+    **kwargs
+        See :ref:`NumPy ufunc docs <ufuncs.kwargs>` for other keyword arguments.
+
+    See also
+    --------
+    union
+    coverage_union
+    disjoint_subset_union_all
+
+    Examples
+    --------
+    >>> from shapely import normalize, Polygon
+    >>> polygon_1 = Polygon([(0, 0), (0, 1), (1, 1), (1, 0), (0, 0)])
+    >>> polygon_2 = Polygon([(1, 0), (1, 1), (2, 1), (2, 0), (1, 0)])
+    >>> normalize(disjoint_subset_union(polygon_1, polygon_2))
+    <POLYGON ((0 0, 0 1, 1 1, 2 1, 2 0, 1 0, 0 0))>
+
+    Union with None returns same polygon:
+
+    >>> normalize(disjoint_subset_union(polygon_1, None))
+    <POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))>
+    """
+    if (isinstance(a, Geometry) or a is None) and (
+        isinstance(b, Geometry) or b is None
+    ):
+        pass
+    elif isinstance(a, Geometry) or a is None:
+        a = np.full_like(b, a)
+    elif isinstance(b, Geometry) or b is None:
+        b = np.full_like(a, b)
+    elif len(a) != len(b):
+        raise ValueError("Arrays a and b must have the same length")
```

Comment:
The logic looks good here (to mimic broadcasting), just wondering if this is already covered by the standard tests we have? (did you have to add this to get the tests passing?)

Reply:
Ah, and it might be that this can be simplified as `a, b = np.broadcast_arrays(a, b)`

Reply:
I tried `broadcast_array` before writing this and it does not work as we need it.  Some of the tests fail with `TypeError: One of the arguments is of incorrect type. Please provide only Geometry objects.`

And yes, I had to include this to make the tests green, so it is properly covering by existing tests.